src/ray/gcs/subscription_executor.cc

```diff
@@ -72,43 +100,50 @@ template <typename ID, typename Data, typename Table>
 Status SubscriptionExecutor<ID, Data, Table>::AsyncSubscribe(
     const ClientID &client_id, const ID &id, const SubscribeCallback<ID, Data> &subscribe,
     const StatusCallback &done) {
-  Status status = AsyncSubscribe(client_id, nullptr, nullptr);
-  if (!status.ok()) {
-    return status;
-  }
-
-  auto on_done = [this, done, id](Status status) {
-    if (!status.ok()) {
-      std::lock_guard<std::mutex> lock(mutex_);
-      id_to_callback_map_.erase(id);
-    }
-    if (done != nullptr) {
-      done(status);
-    }
-  };
 
   {
-    std::lock_guard<std::mutex> lock(mutex_);
+    std::unique_lock<std::mutex> lock(mutex_);
     const auto it = id_to_callback_map_.find(id);
     if (it != id_to_callback_map_.end()) {
       RAY_LOG(DEBUG) << "Duplicate subscription to id " << id << " client_id "
                      << client_id;
       return Status::Invalid("Duplicate subscription to element!");
     }
-    status = table_.RequestNotifications(JobID::Nil(), id, client_id, on_done);
-    if (status.ok()) {
-      id_to_callback_map_[id] = subscribe;
-    }
   }
 
-  return status;
+  // NOTE(zhijunfu): `Subscribe` and other operations use different redis contexts,
+  // thus we need to call `RequestNotifications` in the Subscribe callback to ensure
+  // it's processed after the `Subscribe` request. Otherwise if `RequestNotifications`
+  // is processed first we will miss the initial notification.
+  auto on_subscribe_done = [this, client_id, id, subscribe, done](Status status) {
+    auto on_request_notification_done = [this, done, id](Status status) {
+      if (!status.ok()) {
+        std::unique_lock<std::mutex> lock(mutex_);
+        id_to_callback_map_.erase(id);
+      }
+      if (done != nullptr) {
+        done(status);
+      }
+    };
+
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      status = table_.RequestNotifications(JobID::Nil(), id, client_id,
```

Comment:
Maybe call `RequestNotifications(...)`  outside of lock?

Reply:
that seems possible

Reply:
but it doesn't hurt perf here since the `table_` is also locked